report/topology.go

```diff
@@ -311,6 +331,119 @@ func (s StringSet) Copy() StringSet {
 	return result
 }
 
+// Metrics is a string->metric map.
+type Metrics map[string]Metric
+
+// Merge merges two sets maps into a fresh set, performing set-union merges as
+// appropriate.
+func (m Metrics) Merge(other Metrics) Metrics {
+	result := m.Copy()
+	for k, v := range other {
+		result[k] = result[k].Merge(v)
+	}
+	return result
+}
+
+// Copy returns a value copy of the sets map.
+func (m Metrics) Copy() Metrics {
+	result := Metrics{}
+	for k, v := range m {
+		result[k] = v.Copy()
+	}
+	return result
+}
+
+// Metric is a list of timeseries data. Clients must use the Add
+// method to add values.
+type Metric struct {
+	Samples []Sample  `json:"samples"`
+	Min     float64   `json:"min"`
+	Max     float64   `json:"max"`
+	First   time.Time `json:"first"`
+	Last    time.Time `json:"last"`
+}
```

Comment:
Could we make this type immutable?  I think we just need to make the Metric returned by Add a copy.  And it we swap the sample list of a ps.List, arranged in reverse order, then we can share the list of sample between copies too.